### ISSUE TITLE
[SDL2] [Wayland] Power-cycling monitor causes SDL2 programs to crash (cause is known and described)

### ISSUE THREAD

**Issue by vanfanel** [libsdl-org/SDL#10075]
Hi there,

In WLRoots-based Wayland compositors (Wayfire, for example) SDL2 programs crash when monitor is turned ON after being turned OFF.

This is a crash log with a debug version of SDL2 (happens with any SDL2 program: SDLPop is used as an example):

```
Thread 1 "prince.exe" received signal SIGSEGV, Segmentation fault.
0x00007ffff7f7cde6 in prepare_zombie (proxy=0x55555567c1e0)
    at ../src/wayland-client.c:443
443		for (i = 0; i < interface->event_count; i++) {
(gdb) bt
#0  0x00007ffff7f7cde6 in prepare_zombie (proxy=0x55555567c1e0) at ../src/wayland-client.c:443
#1  0x00007ffff7f7d0ab in proxy_destroy (proxy=0x55555567c1e0) at ../src/wayland-client.c:570
#2  0x00007ffff7f7d189 in wl_proxy_destroy_caller_locks (proxy=0x55555567c1e0)
    at ../src/wayland-client.c:598
#3  0x00007ffff7f7d1c2 in wl_proxy_destroy (proxy=0x55555567c1e0) at ../src/wayland-client.c:621
#4  0x00007ffff7de995e in wl_output_destroy (wl_output=0x55555567c1e0)
    at gen/wayland-client-protocol.h:5726
#5  0x00007ffff7deb7de in Wayland_free_display (d=0x555555623af0, id=39)
    at /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:758
#6  0x00007ffff7debf11 in display_remove_global (data=0x555555623af0, registry=0x555555624d80, 
    id=39) at /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:887
#7  0x00007ffff7f70f7a in ?? () from /lib/x86_64-linux-gnu/libffi.so.8
#8  0x00007ffff7f7040e in ?? () from /lib/x86_64-linux-gnu/libffi.so.8
#9  0x00007ffff7f70b0d in ffi_call () from /lib/x86_64-linux-gnu/libffi.so.8
#10 0x00007ffff7f81f50 in wl_closure_invoke (closure=0x55555682e0e0, flags=1, 
    target=0x555555624d80, opcode=1, data=0x555555623af0) at ../src/connection.c:1228
#11 0x00007ffff7f7eabc in dispatch_event (display=0x55555561f850, queue=0x55555561f948)
    at ../src/wayland-client.c:1670
#12 0x00007ffff7f7eda2 in dispatch_queue (display=0x55555561f850, queue=0x55555561f948)
    at ../src/wayland-client.c:1816
#13 0x00007ffff7f7f055 in wl_display_dispatch_queue_pending (display=0x55555561f850, 
    queue=0x55555561f948) at ../src/wayland-client.c:2058
#14 0x00007ffff7f7f0bd in wl_display_dispatch_pending (display=0x55555561f850)
    at ../src/wayland-client.c:2121
#15 0x00007ffff7de212d in Wayland_PumpEvents (_this=0x555555623d30)
    at /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandevents.c:380
#16 0x00007ffff7c58616 in SDL_PumpEventsInternal (push_sentinel=SDL_TRUE)
--Type <RET> for more, q to quit, c to continue without paging--
    at /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:918
#17 0x00007ffff7c589ef in SDL_WaitEventTimeout_REAL (event=0x7fffffffe0d0, timeout=0)
    at /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:1093
#18 0x00007ffff7c586e6 in SDL_PollEvent_REAL (event=0x7fffffffe0d0)
    at /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:960
#19 0x00007ffff7c4aa90 in SDL_PollEvent (a=0x7fffffffe0d0)
    at /root/src/sdl/SDL2-2.30.4/src/dynapi/SDL_dynapi_procs.h:156
#20 0x0000555555579db0 in process_events ()
#21 0x000055555557c3ba in do_wait ()
#22 0x000055555555ea48 in show_title ()
#23 0x000055555555d827 in start_game ()
#24 0x000055555555f9a3 in pop_main ()
#25 0x000055555555a9d6 in main ()

```

**THE CAUSE**

It's a "double free" bug. Looking at this here:

https://github.com/libsdl-org/SDL/blob/SDL2/src/video/wayland/SDL_waylandvideo.c#L764

This frees the driverdata:

https://github.com/libsdl-org/SDL/blob/SDL2/src/video/SDL_video.c#L681

Which is the same as the data pointer here:

https://github.com/libsdl-org/SDL/blob/SDL2/src/video/wayland/SDL_waylandvideo.c#L741

Which is used after the free call:

https://github.com/libsdl-org/SDL/blob/SDL2/src/video/wayland/SDL_waylandvideo.c#L765-L769

This is an example program (any SDL2 program will do) running with address sanitizer, confirming the cause of the issue:

```
~/prince$ LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libasan.so.8 prince.exe 
=================================================================
==13225==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000191e50 at pc 0x74e41c025510 bp 0x7fff094664b0 sp 0x7fff094664a8
READ of size 8 at 0x611000191e50 thread T0
    #0 0x74e41c02550f in Wayland_free_display /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:757
    #1 0x74e41c0261c3 in display_remove_global /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:892
    #2 0x74e41b7e8f79  (/lib/x86_64-linux-gnu/libffi.so.8+0x6f79)
    #3 0x74e41b7e840d  (/lib/x86_64-linux-gnu/libffi.so.8+0x640d)
    #4 0x74e41b7e8b0c in ffi_call (/lib/x86_64-linux-gnu/libffi.so.8+0x6b0c)
    #5 0x74e41b7f9f4f in wl_closure_invoke ../src/connection.c:1228
    #6 0x74e41b7f6abb in dispatch_event ../src/wayland-client.c:1670
    #7 0x74e41b7f6da1 in dispatch_queue ../src/wayland-client.c:1816
    #8 0x74e41b7f7054 in wl_display_dispatch_queue_pending ../src/wayland-client.c:2058
    #9 0x74e41b7f70bc in wl_display_dispatch_pending ../src/wayland-client.c:2121
    #10 0x74e41c00b57f in Wayland_PumpEvents /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandevents.c:380
    #11 0x74e41bc88d1f in SDL_PumpEventsInternal /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:918
    #12 0x74e41bc893b0 in SDL_WaitEventTimeout_REAL /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:1093
    #13 0x74e41bc88e76 in SDL_PollEvent_REAL /root/src/sdl/SDL2-2.30.4/src/events/SDL_events.c:960
    #14 0x74e41bc76026 in SDL_PollEvent /root/src/sdl/SDL2-2.30.4/src/dynapi/SDL_dynapi_procs.h:156
    #15 0x5b42e42b70e7 in process_events /root/src/SDLPoP-1.23/src/seg009.c:3353
    #16 0x5b42e42baa38 in idle /root/src/SDLPoP-1.23/src/seg009.c:3698
    #17 0x5b42e425ee3c in show_title /root/src/SDLPoP-1.23/src/seg000.c:1908
    #18 0x5b42e425a954 in start_game /root/src/SDLPoP-1.23/src/seg000.c:235
    #19 0x5b42e426163e in pop_main /root/src/SDLPoP-1.23/src/seg000.c:149
    #20 0x5b42e4252e62 in main /root/src/SDLPoP-1.23/src/main.c:27
    #21 0x74e41b846249  (/lib/x86_64-linux-gnu/libc.so.6+0x27249)
    #22 0x74e41b846304 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27304)
    #23 0x5b42e4253270 in _start (/root/prince/prince.exe+0x2c270)

0x611000191e50 is located 16 bytes inside of 224-byte region [0x611000191e40,0x611000191f20)
freed by thread T0 here:
    #0 0x74e41c6b76a8 in __interceptor_free ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x74e41bdbaa1a in real_free /root/src/sdl/SDL2-2.30.4/src/stdlib/SDL_malloc.c:5199
    #2 0x74e41bdbae48 in SDL_free_REAL /root/src/sdl/SDL2-2.30.4/src/stdlib/SDL_malloc.c:5339
    #3 0x74e41bf3407d in SDL_DelVideoDisplay /root/src/sdl/SDL2-2.30.4/src/video/SDL_video.c:676
    #4 0x74e41c0254ea in Wayland_free_display /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:756
    #5 0x74e41c0261c3 in display_remove_global /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:892
    #6 0x74e41b7e8f79  (/lib/x86_64-linux-gnu/libffi.so.8+0x6f79)

previously allocated by thread T0 here:
    #0 0x74e41c6b89cf in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x74e41bdba9b6 in real_malloc /root/src/sdl/SDL2-2.30.4/src/stdlib/SDL_malloc.c:5196
    #2 0x74e41bdbad2d in SDL_malloc_REAL /root/src/sdl/SDL2-2.30.4/src/stdlib/SDL_malloc.c:5295
    #3 0x74e41c024f61 in Wayland_add_display /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:702
    #4 0x74e41c0258e5 in display_handle_global /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:837
    #5 0x74e41b7e8f79  (/lib/x86_64-linux-gnu/libffi.so.8+0x6f79)

SUMMARY: AddressSanitizer: heap-use-after-free /root/src/sdl/SDL2-2.30.4/src/video/wayland/SDL_waylandvideo.c:757 in Wayland_free_display
Shadow bytes around the buggy address:
  0x0c228002a370: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c228002a380: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228002a390: fd fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c228002a3a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228002a3b0: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
=>0x0c228002a3c0: fa fa fa fa fa fa fa fa fd fd[fd]fd fd fd fd fd
  0x0c228002a3d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228002a3e0: fd fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c228002a3f0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c228002a400: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa fa
  0x0c228002a410: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==13225==ABORTING

```

**Comment by Kontrabant** [comment]
Fixed in SDL2 main and 2.30.x via #10122

**Comment by vanfanel** [comment]
Yes! Fix is perfect! The bug is gone :)
Many thanks for looking into this!
